File: addon/LevelUp.lua
local _, WS = ...

-- Level-up boon generation and application: weapon ranks, new weapons,
-- passives, evolutions, and the "everything is maxed" bread fallback.
-- Choice `type` doubles as the quality color key in UI_Overlays.
WS.LevelUp = {}
local LevelUp = WS.LevelUp

-- Weighted sampling without replacement from a candidates array.
local function TakeWeighted(candidates)
    if #candidates == 0 then return nil end
    local pick = WS.WeightedPick(candidates)
    for i = 1, #candidates do
        if candidates[i] == pick then
            candidates[i] = candidates[#candidates]
            candidates[#candidates] = nil
            break
        end
    end
    return pick
end

function LevelUp:BuildChoices(player)
    local candidates = {}

    -- Owned weapons: rank-ups, or the evolution once rank 8 is reached and
    -- the paired passive has been learned.
    for i = 1, #player.weapons do
        local weapon = player.weapons[i]
        local data = weapon.data
        if not player.banished[weapon.id] then
            if weapon.level < WS.WEAPON_MAX_LEVEL then
                candidates[#candidates + 1] = {
                    type = "weapon_rank", id = weapon.id, icon = data.icon, weight = 3,
                    name = data.name,
                    description = data.description,
                    note = ("Rank %d  >  %d"):format(weapon.level, weapon.level + 1),
                }
            elseif not weapon.evolved and (player.upgradeLevels[data.evolvePairing] or 0) > 0 then
                candidates[#candidates + 1] = {
                    type = "evolve", id = weapon.id, icon = data.evolveIcon, weight = 6,
                    name = data.evolveName,
                    description = data.evolveDescription,
                    note = "EVOLUTION - " .. data.name .. " transformed",
                }
            end
        end
    end

    -- Weapon unions: both source weapons must be FULLY EVOLVED to merge into one
    -- super-weapon (freeing a slot). A forged union is recorded permanently so it
    -- can never be offered - or merged - a second time.
    for i = 1, #WS.Unions do
        local recipe = WS.Unions[i]
        local a, b = recipe.from[1], recipe.from[2]
        local wa = WS.Player:GetWeapon(player, a)
        local wb = WS.Player:GetWeapon(player, b)
        if not (player.unionsForged and player.unionsForged[recipe.result])
            and not player.weaponLevels[recipe.result]
            and wa and wa.evolved and wb and wb.evolved then
            local data = WS.Weapons[recipe.result]
            candidates[#candidates + 1] = {
                type = "union", id = recipe.result, icon = data.icon, weight = 10,
                unionFrom = recipe.from,
                name = data.name,
                description = data.description,
                note = ("UNION - merges evolved %s + %s"):format(WS.Weapons[a].name, WS.Weapons[b].name),
            }
        end
    end

    -- New weapons, while there are free slots.
    if #player.weapons < WS.MAX_WEAPONS then
        for i = 1, #WS.WeaponOrder do
            local id = WS.WeaponOrder[i]
            if not player.weaponLevels[id] and not player.banished[id] then
                local data = WS.Weapons[id]
                candidates[#candidates + 1] = {
                    type = "new_weapon", id = id, icon = data.icon, weight = 2,
                    name = data.name,
                    description = data.description,
                    note = "New Weapon  -  pairs with " .. WS.Upgrades[data.evolvePairing].name,
                }
            end
        end
    end

    -- Passive boons.
    for i = 1, #WS.UpgradeOrder do
        local id = WS.UpgradeOrder[i]
        local upgrade = WS.Upgrades[id]
        local rank = player.upgradeLevels[id] or 0
        -- An upgrade may declare `offer(player)` to stay out of the pool until it
        -- would actually do something (e.g. Desecration needs healing to convert).
        local offerable = (not upgrade.offer) or upgrade.offer(player)
        if rank < upgrade.max and not player.banished[id] and offerable then
            candidates[#candidates + 1] = {
                type = "stat", id = id, icon = upgrade.icon, weight = 2,
                name = upgrade.name,
                description = upgrade.description,
                note = ("Rank %d / %d"):format(rank + 1, upgrade.max),
            }
        end
    end

    -- Deep-run fallback breads: a finished build eats bread of varying quality.
    local breads = {
        { type = "bread", id = "crust", icon = "Interface\\Icons\\INV_Misc_Food_11",
          name = "Crust of Bread", description = "A quick bite. Restores 30 health.",
          heal = 30, gold = 10, note = "Better than nothing" },
        { type = "bread", id = "loaf", icon = "Interface\\Icons\\INV_Misc_Food_11",
          name = "Hearty Loaf", description = "A hero's supper. Restores 70 health and a little gold.",
          heal = 70, gold = 25, note = "The innkeeper's finest" },
        { type = "bread", id = "feast", icon = "Interface\\Icons\\INV_Misc_Food_15",
          name = "Traveler's Feast", description = "A full spread. Restores 130 health.",
          heal = 130, gold = 20, note = "Fit for a champion" },
        { type = "bread", id = "bakers_dozen", icon = "Interface\\Icons\\INV_Misc_Food_11",
          name = "Baker's Dozen", description = "Modest food, a fat coin purse. Restores 40 health, +75 gold.",
          heal = 40, gold = 75, note = "Mostly for the gold" },
    }

    -- Fill all three slots from the weighted candidate pool FIRST (new/ranked
    -- weapons, evolutions, unions, passives). A real build never loses a choice to
    -- Limit Break - you always get three genuine picks while any remain.
    local choices = {}
    for _ = 1, 3 do
        local pick = TakeWeighted(candidates)
        if pick then choices[#choices + 1] = pick end
    end

    -- Only once there's nothing left to take (the pool couldn't fill three slots)
    -- does Limit Break appear - permanent weapon damage, repeatable - taking ONE of
    -- the leftover slots. It never crowds out a real upgrade, and it no longer needs
    -- endless/victory: any complete build (even mid normal run) unlocks it.
    if #choices < 3 then
        local pct = WS.floor((WS.Config.limitBreakDamage or 0.08) * 100 + 0.5)
        local template = WS.Config.limitBreakDesc or "Break past your limits: +%d%% weapon damage, forever."
        -- Fill a "%d" in the (editable) template with the current +%; tolerate a
        -- template that has no format specifier.
        local ok, desc = pcall(string.format, template, pct)
        choices[#choices + 1] = {
            type = "limit_break", id = "limit_break", icon = "Interface\\Icons\\Ability_Warrior_Rampage",
            name = WS.Config.limitBreakName or "Limit Break",
            description = ok and desc or template,
            note = ("Limit Breaks: %d"):format(player.limitBreaks or 0),
        }
    end

    -- Any still-empty slots get distinct bread variants (no repeats).
    local pool = {}
    for i = 1, #breads do pool[i] = breads[i] end
    while #choices < 3 and #pool > 0 do
        local index = WS.random(1, #pool)
        choices[#choices + 1] = pool[index]
        pool[index] = pool[#pool]
        pool[#pool] = nil
    end
    return choices
end

function LevelUp:IsEndless()
    local run = WS.Game.run
    return run and (run.mode == "endless" or run.victorious)
end

-- Three random blessings, excluding any the survivor has already taken (so the
-- second offering at 10:00 can never re-offer your run-start blessing).
function LevelUp:BuildBlessingChoices(player)
    local ids = {}
    for i = 1, #WS.BlessingOrder do
        local id = WS.BlessingOrder[i]
        if not (player.blessingsTaken and player.blessingsTaken[id]) then
            ids[#ids + 1] = id
        end
    end
    local choices = {}
    for _ = 1, WS.min(3, #ids) do
        local index = WS.random(1, #ids)
        local id = ids[index]
        ids[index] = ids[#ids]
        ids[#ids] = nil
        local blessing = WS.Blessings[id]
        choices[#choices + 1] = {
            type = "blessing", id = id, icon = blessing.icon,
            name = blessing.name,
            description = WS.FormatTokens(blessing.description, blessing),
            note = "Blessing - permanent for this run",
        }
    end
    return choices
end

function LevelUp:Apply(player, choice)
    if choice.type == "blessing" then
        local blessing = WS.Blessings[choice.id]
        blessing.apply(player, blessing)
        player.blessingName = blessing.name
        player.blessingNames = player.blessingNames or {}
        player.blessingNames[#player.blessingNames + 1] = blessing.name -- keep ALL (start + 10:00)
        player.blessingsTaken = player.blessingsTaken or {}
        player.blessingsTaken[choice.id] = true -- never re-offer a taken blessing
        WS.Audio:Play("evolve")
        WS.Game:Announce(blessing.name, "The blessing takes hold.", 2.5)
        return
    end

    if choice.type == "union" then
        -- Consume the two source weapons, then forge the union super-weapon
        -- at max rank.
        for i = 1, #choice.unionFrom do
            WS.Player:RemoveWeapon(player, choice.unionFrom[i])
        end
        local weapon = WS.Player:AddWeapon(player, choice.id)
        if weapon then
            weapon.level = WS.WEAPON_MAX_LEVEL
            player.weaponLevels[choice.id] = WS.WEAPON_MAX_LEVEL
        end
        player.unionsForged = player.unionsForged or {}
        player.unionsForged[choice.id] = true -- forged once, never again
        WoWSurvivorsDB.statistics.unions = (WoWSurvivorsDB.statistics.unions or 0) + 1
        WS.Audio:Play("evolve")
        WS.UI:Shake(6, 0.5)
        WS.Game:Announce(choice.name .. "!", "Two weapons become one.", 3.0)
        WS.UI:RefreshWeaponStrip(player)

    elseif choice.type == "new_weapon" then
        WS.Player:AddWeapon(player, choice.id)
        WS.Audio:Play("chest") -- a new weapon deserves a loot sting

    elseif choice.type == "weapon_rank" then
        WS.Player:LevelWeapon(player, choice.id)

    elseif choice.type == "evolve" then
        local weapon = WS.Player:GetWeapon(player, choice.id)
        if weapon and not weapon.evolved then
            weapon.evolved = true
            WoWSurvivorsDB.statistics.evolutions = WoWSurvivorsDB.statistics.evolutions + 1
            WS.Audio:Play("evolve")
            WS.UI:Shake(5, 0.4)
            WS.Game:Announce(weapon.data.evolveName .. "!", "Your weapon has evolved", 3.0)
            WS.UI:RefreshWeaponStrip(player)
        end

    elseif choice.type == "stat" then
        local upgrade = WS.Upgrades[choice.id]
        upgrade.apply(player, upgrade)
        player.upgradeLevels[choice.id] = (player.upgradeLevels[choice.id] or 0) + 1
        WS.UI:RefreshPassiveStrip(player)

    elseif choice.type == "bread" then
        WS.Player:Heal(player, choice.heal or 60, "food")
        if choice.gold and choice.gold > 0 then
            WS.Game:AddGold(WS.floor(choice.gold * WS.Game.run.goldMult), player.x, player.y)
        end
        WS.Audio:Play("potion")

    elseif choice.type == "limit_break" then
        player.damageMultiplier = player.damageMultiplier + (WS.Config.limitBreakDamage or 0.08)
        player.limitBreaks = (player.limitBreaks or 0) + 1
        WS.Audio:Play("evolve")
        WS.FloatingText:Notice(player.x, player.y, "LIMIT BREAK!", 1, 0.5, 0.2)
    end
end

-- Removes the boon (and its future offers) from this run's pool.
function LevelUp:Banish(player, choice)
    if choice.type == "bread" or choice.type == "blessing"
        or choice.type == "limit_break" or choice.type == "union" then
        return false
    end
    player.banished[choice.id] = true
    return true
end

File: src/game/levelup.js
/* Level-up boon generation and application: weapon ranks, new weapons,
 * evolutions, unions, passives, the Breaking Point, and the deep-run bread
 * fallback. `type` doubles as the card's quality key in the UI. */
'use strict';
(function (WS) {

  const LevelUp = {};

  const BREADS = [
    {
      type: 'bread', id: 'crust', art: 'leaf', name: 'Crust of Bread',
      description: 'A quick bite. Restores 30 health.', heal: 30, gold: 10,
      note: 'Better than nothing',
    },
    {
      type: 'bread', id: 'loaf', art: 'leaf', name: 'Hearty Loaf',
      description: "A hero's supper. Restores 70 health and a little gold.", heal: 70, gold: 25,
      note: "The innkeeper's finest",
    },
    {
      type: 'bread', id: 'feast', art: 'heart', name: "Traveler's Feast",
      description: 'A full spread. Restores 130 health.', heal: 130, gold: 20,
      note: 'Fit for a champion',
    },
    {
      type: 'bread', id: 'bakers_dozen', art: 'coin', name: "Baker's Dozen",
      description: 'Modest food, a fat coin purse. Restores 40 health, +75 gold.', heal: 40, gold: 75,
      note: 'Mostly for the gold',
    },
  ];

  /* ------------------------------------------------- what this will do --
   * WHAT A CARD REACTS WITH, said on the card.
   *
   * Everything in this game is built around things combining - a weapon and
   * a passive become an evolution, two weapons become a discovery, two
   * evolutions become a union - and a card said none of it. It gave a name, a
   * rank and a sentence, and the entire combination layer was something a
   * player either read in the Codex afterwards or never found. A survivor
   * holding Cinderfall, offered Rimeshard, was being offered Frostfire and
   * had no way to know.
   *
   * So a card carries its reactions: what it completes NOW, and what it is
   * waiting on. `ready` is the difference - a reaction the player can have by
   * taking this card reads lit, one that needs something they do not have yet
   * reads as a note. Both are worth showing; only one is a decision.
   */
  function reactionsFor(p, id, kind, level) {
    const out = [];
    const d = WS.Weapons[id];
    if (!d) return out;
    const lvl = level === undefined ? (p.weaponLevels[id] || 0) : level;

    // A discovery, with something already carried.
    for (const cid of WS.ComboOrder) {
      const c = WS.Combos[cid];
      if (c.weapons.indexOf(id) < 0) continue;
      const other = c.weapons[0] === id ? c.weapons[1] : c.weapons[0];
      if (p.combosActive[cid]) continue;
      const held = !!WS.Player.getWeapon(p, other);
      out.push({ kind: 'discovery', ready: held && kind === 'new_weapon',
        text: held ? 'Discovery: ' + c.name
          : c.name + ' needs ' + (WS.Weapons[other] ? WS.Weapons[other].name : other) });
    }

    // The evolution this weapon is walking toward.
    if (d.evolvePairing && WS.Upgrades[d.evolvePairing]) {
      const passive = WS.Upgrades[d.evolvePairing].name;
      const has = (p.upgradeLevels[d.evolvePairing] || 0) > 0;
      const maxed = lvl >= WS.WEAPON_MAX_LEVEL;
      if (kind !== 'evolve') {
        out.push({ kind: 'evolve', ready: has && maxed,
          text: has
            ? (maxed ? 'Ready to evolve: ' + d.evolveName
              : 'Evolves into ' + d.evolveName + ' at rank ' + WS.WEAPON_MAX_LEVEL)
            : 'Evolves with ' + passive });
      }
    }

    // And the union two evolutions can become.
    for (const u of (WS.Unions || [])) {
      if (u.from.indexOf(id) < 0) continue;
      const other = u.from[0] === id ? u.from[1] : u.from[0];
      const ow = WS.Player.getWeapon(p, other);
      const res = WS.Weapons[u.result];
      out.push({ kind: 'union', ready: !!(ow && ow.evolved),
        text: (res ? res.name : 'Union') + ' once '
          + (WS.Weapons[other] ? WS.Weapons[other].name : other) + ' evolves' });
    }
    /* Two at most, and what is achievable first. A card that lists every
       future it could have is a wall of text, not an aid. */
    out.sort((a, b) => (b.ready ? 1 : 0) - (a.ready ? 1 : 0));
    return out.slice(0, 2);
  }
  LevelUp.reactionsFor = reactionsFor;

  /* What a PASSIVE reacts with: the weapons it evolves. A passive card said
   * what it does to your numbers and nothing about the one thing that makes
   * most of them worth taking - that Ferocity is what turns a rank-8
   * Knifestorm into Steel Flurry. Weapons you carry come first, and one that
   * is already at its last rank is the reaction you can have by taking it. */
  function passiveReactions(p, id, all) {
    const out = [];
    const has = (p.upgradeLevels[id] || 0) > 0;
    for (const wid of WS.WeaponOrder) {
      const d = WS.Weapons[wid];
      if (!d || d.evolvePairing !== id) continue;
      const w = WS.Player.getWeapon(p, wid);
      if (w && w.evolved) continue;
      if (w) {
        const maxed = w.level >= WS.WEAPON_MAX_LEVEL;
        out.push({ kind: 'evolve', carried: true, ready: maxed && !has, weapon: wid,
          text: maxed ? (has ? 'Evolving ' : 'Evolves your ') + d.name + ' into ' + d.evolveName + ' now'
            : 'Evolves your ' + d.name + ' into ' + d.evolveName + ' at rank ' + WS.WEAPON_MAX_LEVEL });
      } else if (all) {
        out.push({ kind: 'evolve', carried: false, ready: false, weapon: wid,
          text: 'Evolves ' + d.name + ' into ' + d.evolveName });
      }
    }
    out.sort((a, b) => (b.ready ? 2 : b.carried ? 1 : 0) - (a.ready ? 2 : a.carried ? 1 : 0));
    return out;
  }
  LevelUp.passiveReactions = passiveReactions;

  LevelUp.buildChoices = function (p) {
    const candidates = [];

    // Owned weapons: a rank-up, or the evolution once rank 8 is reached and
    // the paired passive has been learned.
    for (const w of p.weapons) {
      const d = w.data;
      if (p.banished[w.id]) continue;
      if (w.level < WS.WEAPON_MAX_LEVEL) {
        candidates.push({
          type: 'weapon_rank', id: w.id, art: d.art, school: d.school, weight: 3,
          name: d.name, description: WS.template(d.description, d),
          note: `Rank ${w.level} to ${w.level + 1}`,
          rank: w.level + 1, maxRank: WS.WEAPON_MAX_LEVEL,
          reacts: reactionsFor(p, w.id, 'weapon_rank', w.level + 1),
        });
      } else if (!w.evolved && d.evolvePairing && (p.upgradeLevels[d.evolvePairing] || 0) > 0) {
        candidates.push({
          type: 'evolve', id: w.id, art: d.art, school: d.school, weight: 6,
          name: d.evolveName, description: d.evolveDescription,
          rank: WS.WEAPON_MAX_LEVEL, maxRank: WS.WEAPON_MAX_LEVEL,
          reacts: reactionsFor(p, w.id, 'evolve', WS.WEAPON_MAX_LEVEL),
          note: 'Evolution · ' + d.name + ', transformed',
        });
      }
    }

    // Unions: both sources fully evolved merge into one super-weapon, freeing
    // a slot. A forged union is recorded so it can never be offered again.
    for (const recipe of WS.Unions) {
      const [a, b] = recipe.from;
      const wa = WS.Player.getWeapon(p, a);
      const wb = WS.Player.getWeapon(p, b);
      if (!p.unionsForged[recipe.result] && !p.weaponLevels[recipe.result]
        && wa && wa.evolved && wb && wb.evolved) {
        const d = WS.Weapons[recipe.result];
        candidates.push({
          type: 'union', id: recipe.result, art: d.art, school: d.school, weight: 10,
          unionFrom: recipe.from, name: d.name, description: WS.template(d.description, d),
          note: `Union · ${WS.Weapons[a].name} and ${WS.Weapons[b].name}, made one`,
        });
      }
    }

    // New weapons, while there are free slots.
    if (p.weapons.length < WS.MAX_WEAPONS) {
      for (const id of WS.WeaponOrder) {
        if (p.weaponLevels[id] || p.banished[id]) continue;
        const d = WS.Weapons[id];
        candidates.push({
          type: 'new_weapon', id, art: d.art, school: d.school, weight: 2,
          name: d.name, description: WS.template(d.description, d),
          rank: 1, maxRank: WS.WEAPON_MAX_LEVEL,
          reacts: reactionsFor(p, id, 'new_weapon', 1),
          note: 'New weapon · pairs with ' + WS.Upgrades[d.evolvePairing].name,
        });
      }
    }

    // Passive boons.
    for (const id of WS.UpgradeOrder) {
      const up = WS.Upgrades[id];
      const rank = p.upgradeLevels[id] || 0;
      const offerable = !up.offer || up.offer(p);
      if (rank < up.max && !p.banished[id] && offerable) {
        candidates.push({
          type: 'stat', id, art: up.art, quality: up.quality, weight: 2,
          name: up.name, description: WS.template(up.description, up),
          detail: WS.template(up.detail, up),
          rank: rank + 1, maxRank: up.max,
          note: `Rank ${rank + 1} of ${up.max}`,
          reacts: passiveReactions(p, id, false).slice(0, 2),
        });
      }
    }

    // Fill all three slots from the real pool FIRST: a live build never loses
    // a choice to Breaking Point while genuine upgrades remain.
    const choices = [];
    for (let i = 0; i < 3; i++) {
      const pick = WS.takeWeighted(candidates);
      if (pick) choices.push(pick);
    }

    // Only once the pool cannot fill three slots does Breaking Point appear.
    if (choices.length < 3) {
      // From here the HUD offers the auto-take switch under the portrait.
      if (WS.Game && WS.Game.run) WS.Game.run.breakingSeen = true;
      const pct = WS.round(WS.Config.limitBreakDamage * 100);
      choices.push({
        type: 'breaking_point', id: 'breaking_point', art: 'fist', quality: 'legendary',
        name: 'Breaking Point',
        description: `Break past your limits: +${pct}% weapon damage, forever.`,
        note: `Breaking Points: ${p.limitBreaks}`,
      });
    }

    // Any still-empty slots get distinct breads.
    const pool = BREADS.slice();
    while (choices.length < 3 && pool.length) {
      choices.push(pool.splice(WS.randInt(0, pool.length - 1), 1)[0]);
    }
    return choices;
  };

  /** Three random blessings, never re-offering one already taken. A class's
   *  signature blessing (Ruinous Pact for the Ruinseeker, Blood Rite for the
   *  Graveblade) always fills one of the three slots while it is still
   *  available - the draft can still hand either to anyone, but the class
   *  whose identity it is is never unlucky enough to miss it. */
  LevelUp.buildBlessingChoices = function (p) {
    const ids = WS.BlessingOrder.filter((id) => !p.blessingsTaken || !p.blessingsTaken[id]);
    const picks = [];
    const sig = p.character && p.character.signatureBlessing;
    if (sig && ids.includes(sig)) picks.push(ids.splice(ids.indexOf(sig), 1)[0]);
    const slots = WS.min(3, ids.length + picks.length);
    while (picks.length < slots) picks.push(ids.splice(WS.randInt(0, ids.length - 1), 1)[0]);
    return picks.map((id) => {
      const b = WS.Blessings[id];
      return {
        type: 'blessing', id, art: b.art, quality: b.quality || 'legendary',
        name: b.name,
        description: WS.template(b.description, b),
        note: 'A blessing · yours for the rest of the night',
      };
    });
  };

  LevelUp.apply = function (p, choice) {
    if (choice.type === 'blessing') {
      const b = WS.Blessings[choice.id];
      b.apply(p, b);
      p.blessingNames.push(b.name);
      p.blessingsTaken = p.blessingsTaken || {};
      p.blessingsTaken[choice.id] = true;
      WS.Audio.play('evolve');
      WS.Game.announce(b.name, 'The blessing takes hold.', 2.5, { kind: 'glory' });
      return;
    }

    if (choice.type === 'union') {
      for (const id of choice.unionFrom) WS.Player.removeWeapon(p, id);
      const w = WS.Player.addWeapon(p, choice.id);
      if (w) { w.level = WS.WEAPON_MAX_LEVEL; p.weaponLevels[choice.id] = WS.WEAPON_MAX_LEVEL; }
      p.unionsForged[choice.id] = true;
      WS.Save.stats.unions++;
      WS.Save.db.unions[choice.id] = true;   // for the codex, which wants which
      WS.Audio.play('evolve');
      WS.FX.shake(6, 0.5);
      WS.Game.announce(choice.name + '!', 'Two weapons become one.', 3.0, { kind: 'glory' });

    } else if (choice.type === 'new_weapon') {
      WS.Player.addWeapon(p, choice.id);
      WS.Audio.play('chest');

    } else if (choice.type === 'weapon_rank') {
      WS.Player.levelWeapon(p, choice.id);
      WS.Audio.play('select');

    } else if (choice.type === 'evolve') {
      const w = WS.Player.getWeapon(p, choice.id);
      if (w && !w.evolved) {
        w.evolved = true;
        WS.Save.stats.evolutions++;
        WS.Save.db.evolved[choice.id] = true;   // ditto
        WS.Audio.play('evolve');
        WS.FX.shake(5, 0.4);
        WS.Game.announce(w.data.evolveName + '!', 'Your weapon has grown into something else.', 3.0,
          { kind: 'glory' });
      }

    } else if (choice.type === 'stat') {
      const up = WS.Upgrades[choice.id];
      up.apply(p, up);
      p.upgradeLevels[choice.id] = (p.upgradeLevels[choice.id] || 0) + 1;
      WS.Audio.play('select');

    } else if (choice.type === 'bread') {
      WS.Player.heal(p, choice.heal || 60, 'food');
      if (choice.gold > 0) WS.Game.addGold(WS.floor(choice.gold * WS.Game.run.goldMult), p.x, p.y);
      WS.Audio.play('potion');

    } else if (choice.type === 'breaking_point') {
      p.damageMultiplier += WS.Config.limitBreakDamage;
      p.limitBreaks++;
      WS.Audio.play('evolve');
      WS.FX.notice(p.x, p.y, 'BREAKING POINT!', '#ff8a3c');
    }
    WS.Achievements.check();
  };

  /** Removes a boon (and its future offers) from this run's pool. */
  LevelUp.banish = function (p, choice) {
    if (choice.type === 'bread' || choice.type === 'blessing'
      || choice.type === 'breaking_point' || choice.type === 'union') return false;
    p.banished[choice.id] = true;
    return true;
  };

  WS.LevelUp = LevelUp;

})(window.WS);
